homeassistant/components/media_player/pjlink.py

```diff
@@ -0,0 +1,148 @@
+"""
+Support for controlling projector via the PJLink protocol.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/media_player.pjlink/
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    SUPPORT_TURN_OFF, SUPPORT_TURN_ON, SUPPORT_VOLUME_MUTE,
+    SUPPORT_SELECT_SOURCE, PLATFORM_SCHEMA, MediaPlayerDevice)
+from homeassistant.const import (
+    STATE_OFF, STATE_ON, CONF_HOST,
+    CONF_NAME, CONF_PASSWORD, CONF_PORT)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['pypjlink2==1.2.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ENCODING = 'encoding'
+
+DEFAULT_PORT = 4352
+DEFAULT_ENCODING = 'utf-8'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_ENCODING, default=DEFAULT_ENCODING): cv.string,
+    vol.Optional(CONF_PASSWORD): cv.string,
+})
+
+SUPPORT_PJLINK = SUPPORT_VOLUME_MUTE | \
+    SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_SELECT_SOURCE
+
+KNOWN_HOSTS = []
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the PJLink platform."""
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    name = config.get(CONF_NAME)
+    encoding = config.get(CONF_ENCODING)
+    password = config.get(CONF_PASSWORD)
+    devices = []
+
+    if CONF_HOST in config and host not in KNOWN_HOSTS:
+        device = PjLinkDevice(host, port, name, encoding, password)
+        KNOWN_HOSTS.append("%s:%s" % (host, port))
+        devices.append(device)
+
+    add_devices(devices)
+
+
+class PjLinkDevice(MediaPlayerDevice):
+    """Representation of a PJLink device."""
+
+    def __init__(self, host, port, name, encoding, password):
+        """Iinitialize the PJLink device."""
+        self._host = host
+        self._port = port
+        self._name = name
+        self._encoding = encoding
+        self._muted = False
+        self._pwstate = STATE_OFF
+        self._current_source = None
+        with self.projector() as projector:
```

Comment:
Why do we use the `with... as` here? The `projector` method doesn't look like a context manager.

Reply:
The value returned by `self.projector()` is an instance of `pypjlink.Projector`, using `with ... as` ensures that the resources used by that instance are correctly closed.

Reply:
I don't see that the library source supports that. Am I missing something? This is the source I've been looking at:

https://github.com/gaetano-guerriero/pypjlink/blob/master/pypjlink/projector.py

Reply:
The code in that repo is unmaintained. I created a fork to add some missing features. The source is there https://github.com/benoitlouy/pypjlink/blob/master/pypjlink/projector.py

Reply:
Ah, sorry, I missed the 2 in the name when I was googling the pypi package.